pkg/file/pipeline/builder/builder.go

```diff
@@ -0,0 +1,128 @@
+
+// newPipeline creates a standard pipeline that only hashes content with BMT to create
+// a merkle-tree of hashes that represent the given arbitrary size byte stream. Partial
+// writes are supported. The pipeline flow is: Data -> Feeder -> BMT -> Storage -> HashTrie.
+func newPipeline(ctx context.Context, s storage.Storer, mode storage.ModePut) pipeline.Interface {
+	tw := hashtrie.NewHashTrieWriter(swarm.ChunkSize, swarm.Branches, swarm.HashSize, newShortPipelineFunc(ctx, s, mode))
+	lsw := store.NewStoreWriter(ctx, s, mode, tw)
+	b := bmt.NewBmtWriter(128, lsw)
+	return feeder.NewChunkFeederWriter(swarm.ChunkSize, b)
+}
+
+// newShortPipelineFunc returns a constructor function for an ephemeral hashing pipeline
+// needed by the hashTrieWriter.
+func newShortPipelineFunc(ctx context.Context, s storage.Storer, mode storage.ModePut) func() pipeline.ChainWriter {
+	return func() pipeline.ChainWriter {
+		lsw := store.NewStoreWriter(ctx, s, mode, nil)
+		return bmt.NewBmtWriter(128, lsw)
+	}
+}
+
+// newEncryptionPipeline creates an encryption pipeline that encrypts using CTR, hashes content with BMT to create
+// a merkle-tree of hashes that represent the given arbitrary size byte stream. Partial
+// writes are supported. The pipeline flow is: Data -> Feeder -> Encryption -> BMT -> Storage -> HashTrie.
+// Note that the encryption writer will mutate the data to contain the encrypted span, but the span field
+// with the unencrypted span is preserved.
+func newEncryptionPipeline(ctx context.Context, s storage.Storer, mode storage.ModePut) pipeline.Interface {
+	tw := hashtrie.NewHashTrieWriter(swarm.ChunkSize, 64, swarm.HashSize+encryption.KeyLength, newShortEncryptionPipelineFunc(ctx, s, mode))
+	lsw := store.NewStoreWriter(ctx, s, mode, tw)
+	b := bmt.NewBmtWriter(128, lsw)
+	enc := enc.NewEncryptionWriter(encryption.NewChunkEncrypter(), b)
+	return feeder.NewChunkFeederWriter(swarm.ChunkSize, enc)
+}
+
+// newShortEncryptionPipelineFunc returns a constructor function for an ephemeral hashing pipeline
+// needed by the hashTrieWriter.
+func newShortEncryptionPipelineFunc(ctx context.Context, s storage.Storer, mode storage.ModePut) func() pipeline.ChainWriter {
+	return func() pipeline.ChainWriter {
+		lsw := store.NewStoreWriter(ctx, s, mode, nil)
+		b := bmt.NewBmtWriter(128, lsw)
+		return enc.NewEncryptionWriter(encryption.NewChunkEncrypter(), b)
+	}
+}
+
+// FeedPipeline feeds the pipeline with the given reader until EOF is reached.
+// It returns the cryptographic root hash of the content.
+func FeedPipeline(ctx context.Context, pipeline pipeline.Interface, r io.Reader, dataLength int64) (addr swarm.Address, err error) {
+	var total int64
+	data := make([]byte, swarm.ChunkSize)
+	var eof bool
+	for !eof {
+		c, err := r.Read(data)
+		total += int64(c)
+		if err != nil {
+			if err == io.EOF {
+				if total < dataLength {
+					return swarm.ZeroAddress, fmt.Errorf("pipline short write: read %d out of %d bytes", total+int64(c), dataLength)
```

Comment:
Why doing again `total+int64(c)` in error message ?
Isn't this already done in previous line, right before checking for error.